packages/debugger-extension/src/index.ts

```diff
@@ -356,7 +358,7 @@ const main: JupyterFrontEndPlugin<void> = {
     settingRegistry: ISettingRegistry | null,
     themeManager: IThemeManager | null,
     debuggerSources: IDebugger.ISources | null
-  ): Promise<void> => {
+  ): Promise<IDebugger.ISidebar | void> => {
```

Comment:
I will change the token definition to include `void`.

Reply:
I guess the token type would include `undefined`

Reply:
I have to declare a type that will be used by the token and export it so it can be used by the provider

Reply:
I don't think that creating a type that include `undefined` is a good "out" here because it still imposes the cost of checking whether the debugger sidebar exists or not on *every single consumer* of this plugin.

I think the plugin should *always* return something, whether or not that thing is visible in the UI.

@jtpio, what do you think?

Reply:
> I don't think that creating a type that include `undefined` is a good "out" here because it still imposes the cost of checking whether the debugger sidebar exists or not on _every single consumer_ of this plugin.

Sounds like that would indeed break the pattern of optionals (i.e `if (plugin) { }`).

> I think the plugin should _always_ return something, whether or not that thing is visible in the UI.

Maybe the sidebar plugin (`@jupyterlab/debugger-extension:main`) could then always return the sidebar, and it would be the responsibility of another plugin to add it to the shell (as well as adding the commands), depending on whether a kernel supports debugging or not.

Reply:
> Maybe the sidebar plugin (@jupyterlab/debugger-extension:main) could then always return the sidebar, and it would be the responsibility of another plugin to add it to the shell (as well as adding the commands), depending on whether a kernel supports debugging or not.

I think this suggestion is more in keeping with the plugin architecture and with the semantic meanings of `requires` and `optional` plugin dependencies.

Reply:
To me, in one sense, always returning something feels cleaner for the debugger, but I'll point out that returning `undefined` plays really well with the plugin being optional, as then a consumer gets undefined if the plugin is installed and not activated, just the same as if it is not installed.

If we always return a debugger, it's too bad that a lot of work is done in the debugger extension that is not used if we are just constructing the object without using it, and possibly a *lot* of other third-party work adding things to the debugger when it isn't used. In this case, I think it would make sense to introduce a new flag in the debugger saying whether it is actually active - but then we're right back to all third-party plugins checking that flag (*and* checking it if the plugin exists, since it should probably be considered an optional dependency). For a 3rd party plugin, it seems more natural to just treat it as an optional extension (and have the plugin return undefined if not active), because their logic would be the same if the plugin wasn't installed or if it was installed and not active.

Reply:
If you are writing a plugin `requires` the debugger UI, you should never need to check if it exists. I agree that instead it would be better to never explicitly require it and always treat it as `optional`.

But if the `activate` function is being called, I think it should always return an actual object instead of `undefined`, always. I don't feel strongly enough about this to argue too vehemently, but the pattern smells fishy to me.

This is a breaking API change, so whatever we decide in this PR is going to last until at least 4.0.

Reply:
> but the pattern smells fishy to me.

I agree. I'm not entirely comfortable with it either.

But I find the argument that a consuming plugin just has to check if the token is undefined as pretty compelling too. The two places it breaks down is:

(a) what you point out, that if you require it, it's weird that it isn't there, and
(b) if you ever want to distinguish between the plugin being in the system and not activated, vs not being in the system.

Reply:
It would be *really* good to come up with a good pattern for a plugin that wants to optionally offer itself to the system, so I think it's good to have this discussion and try to come up with something that is elegant.

Reply:
Maybe for now we could follow the same approach as for the status bar?

The status bar can be hidden using the "Show Status Bar" command / menu item. But this doesn't make unavailable to other plugins when it's hidden.

Other plugins can still require the `IStatusBar` token and add items to it when it exists in the system (even though it might not be visible). This is for example the case of the memory indicator:

https://github.com/jupyterlab/jupyterlab/blob/abf1c6349e948375618f597082af7da07286c12a/packages/statusbar-extension/src/index.ts#L330-L352

For the debugger we could think of a similar behavior. The sidebar would always be created, and the `@jupyterlab/debugger-extension:main` plugin would still control its visibility by adding it to the shell or not. At some point we were also thinking about having this kind of control given to the user via the settings editor.

Also the debugger will eventually be displayed by default in the future, and in this case the sidebar will always be available and visible.

Reply:
What if we separate the sidebar creation logic, as suggested by @jtpio , into its own plugin which would provide a factory rather than an instance of the sidebar?
In this case `@jupyterlab/debugger-extension:main` would no longer be a provider and it could do all it checks before calling the factory to create / get default instance. And we would also no longer waist resources to create sidebar that would not be used.

I know this will not resolve the question of how to handle optionally offered plugins. To me it seems that that discussion will require quite a bit of consideration.

Thoughts?

Reply:
> This is a breaking API change, so whatever we decide in this PR is going to last until at least 4.0.

Looking through this, I don't think this PR is backwards-incompatible, i.e., it could be merged after 3.0 for 3.1. There are two major changes:

1. Adding a few functions to the debugger widget - this is just adding functions, not backwards incompatible
2. Having the plugin provide a resource to the system. Since the plugin was providing nothing to the system before, I think providing something is backwards compatible.

We're past the extended deadline for having an RC, and there is enough discussion here that perhaps it is better to consider for a 3.1, especially if the consensus is that these changes are backwards compatible.

Reply:
This came up in our weekly call just now and the consensus was that if we take @jtpio's proposal to follow a similar pattern to the way the status bar works (i.e., it doesn't matter whether it is going to be visible or not, other plugins can rely on its API existing). If we can get this for 3.0 final, let's release it. If it takes longer, then 3.1 is a reasonable option, too.

Reply:
I have separated sidebar creation from the main plugin as recommended by @jtpio .
@afshin @jtpio @jasongrout 
Please take a look and let me know what you think

Reply:
Thanks @mnowacki-b